src/state_manager/mod.rs

```diff
@@ -1315,37 +1106,195 @@ where
             })
     }
 
-    fn chain_rand(&self, blocks: TipsetKeys) -> ChainRand<DB> {
+    fn chain_rand(&self, tipset: Arc<Tipset>) -> ChainRand<DB> {
         ChainRand::new(
             self.chain_config.clone(),
-            blocks,
+            tipset,
             self.cs.clone(),
             self.beacon.clone(),
         )
     }
 }
 
-fn chain_epoch_root<DB>(
-    sm: Arc<StateManager<DB>>,
+/// Messages are transactions that produce new states. The state (usually
```

Comment:
Hmm, don't think we should link to `compute_tipset_state`. That function just calls `apply_block_messages` with appropriate caches. Open to suggestions.

Reply:
It's just for the definition - "Messages are transactions that produce new states" is slightly different from `compute_tipset_state` s documentation. Is one more right than the other?

Reply:
Oh, you mean `compute_tipset_state` states that all messages are a single transaction?

Reply:
I've changed the documentation for `compute_tipset_state` to avoid using the word `transaction`. Its documentation is now a high-level description, and it refers to `apply_block_messages` for details.